File: packages/pdp-geo-eval-agent/src/citation/source-text.ts
import type { EvalNormalizedProduct } from "../types";

/**
 * Assembles the vanilla (pre-generation) PDP source text from a normalized
 * product signal. This is the counterfactual control document for the
 * citation probe: the same probe slot is alternately filled with this text
 * and the generated PDP text, so both variants compete against identical
 * distractors on identical queries.
 *
 * The layout intentionally mirrors a raw PDP dump (title, marketing copy,
 * spec-ish lists) rather than an optimized document.
 */
export function buildVanillaSourceText(product: EvalNormalizedProduct): string {
  const lines: string[] = [];
  lines.push([product.brand, product.name].filter(Boolean).join(" "));
  if (product.category) {
    lines.push(`Category: ${product.category}`);
  }
  if (product.sourceTexts && product.sourceTexts.length > 0) {
    lines.push(...product.sourceTexts);
  }
  if (product.benefits && product.benefits.length > 0) {
    lines.push(`Benefits: ${product.benefits.join(", ")}`);
  }
  if (product.effects && product.effects.length > 0) {
    lines.push(`Effects: ${product.effects.join(", ")}`);
  }
  if (product.ingredients && product.ingredients.length > 0) {
    lines.push(`Key ingredients: ${product.ingredients.join(", ")}`);
  }
  if (product.usage && product.usage.length > 0) {
    lines.push(`How to use: ${product.usage.join(" ")}`);
  }
  if (product.faq && product.faq.length > 0) {
    for (const item of product.faq) {
      lines.push(`Q: ${item.question} A: ${item.answer}`);
    }
  }
  if (product.reviews?.keywords && product.reviews.keywords.length > 0) {
    lines.push(`Review keywords: ${product.reviews.keywords.join(", ")}`);
  }
  return lines.filter((line) => line.trim().length > 0).join("\n");
}

File: packages/pdp-geo-eval-agent/src/pdp_geo_eval_agent/citation/source_text.py
"""Counterfactual vanilla PDP source-text builder."""

from __future__ import annotations

from collections.abc import Mapping, Sequence
from typing import cast


def build_vanilla_source_text(product: object) -> str:
    """Flatten normalized product signals in the legacy raw-PDP order."""
    lines: list[str] = []
    brand = _field(product, "brand")
    name = _field(product, "name")
    identity = " ".join(part for part in (brand, name) if isinstance(part, str) and part)
    lines.append(identity)
    category = _field(product, "category")
    if isinstance(category, str) and category:
        lines.append(f"Category: {category}")
    source_texts = _string_list(_field(product, "sourceTexts", "source_texts"))
    if source_texts:
        lines.extend(source_texts)
    benefits = _string_list(_field(product, "benefits"))
    if benefits:
        lines.append(f"Benefits: {', '.join(benefits)}")
    effects = _string_list(_field(product, "effects"))
    if effects:
        lines.append(f"Effects: {', '.join(effects)}")
    ingredients = _string_list(_field(product, "ingredients"))
    if ingredients:
        lines.append(f"Key ingredients: {', '.join(ingredients)}")
    usage = _string_list(_field(product, "usage"))
    if usage:
        lines.append(f"How to use: {' '.join(usage)}")
    faq = _field(product, "faq")
    if _sequence(faq):
        for item in _sequence(faq):
            question = _field(item, "question")
            answer = _field(item, "answer")
            lines.append(f"Q: {question} A: {answer}")
    reviews = _field(product, "reviews")
    keywords = _string_list(_field(reviews, "keywords")) if reviews is not None else []
    if keywords:
        lines.append(f"Review keywords: {', '.join(keywords)}")
    return "\n".join(line for line in lines if line.strip())


def _string_list(value: object) -> list[str]:
    return [item for item in _sequence(value) if isinstance(item, str)]


def _field(value: object, *names: str) -> object:
    for name in names:
        if isinstance(value, Mapping) and name in cast(Mapping[str, object], value):
            return cast(Mapping[str, object], value)[name]
        if not isinstance(value, Mapping) and hasattr(value, name):
            return cast(object, getattr(value, name))
    return None


def _sequence(value: object) -> list[object]:
    if not isinstance(value, Sequence) or isinstance(value, str):
        return []
    return list(cast(Sequence[object], value))
